### PR TITLE
reconcile: allow atomic reconciliation of release

### PR DESCRIPTION
**Pull request by hiddeco**
This pull requests adds "atomic release" behavior to the internal reconcilers.

In other words, this means that it determines the current state of the Helm release and decides what action is best to take next, until it has reached the desired state or is not allowed to proceed any further. For example, because of it running a successful upgrade, or having run a remediation action.

In addition to this:

- Bits have been added which deal with summarizing the API specific condition types into generic `Ready`, `Stalled` and/or `Progressing`  types.
- The API has been further improved to move methods to the `HelmRelease` instead of `HelmRelease.Spec`.
- The emitting of events is now tested.
- Release names exceeding the maximum limit are truncated.
- Plus a variety of other tiny adjustments, see the individual commits.

With this added, we can move forward to an actual "main" reconciler implementation and further tidying of the UX.

### CODE REVIEW

**Comment by palexster** [comment]
Hi @hiddeco, is this PR related to solve https://github.com/fluxcd/helm-controller/issues/186?

**Comment by hiddeco** [comment]
It is in preparation of, but not solving it by itself.

**Comment by faganihajizada** [comment]
Thanks, @kingdonb 🤝  happy new year! 🍾🥳

File: internal/action/verify.go [hunk 1]
```diff
@@ -33,57 +33,126 @@ import (
...
-	if err := obs.Encode(verifier); err != nil {
+	if err = obs.Encode(verifier); err != nil {
 		// We are expected to be able to encode valid JSON, error out without a
 		// typed error assuming malfunction to signal to e.g. retry.
-		return nil, err
+		return err
 	}
 	if !verifier.Verified() {
-		return nil, ErrReleaseNotObserved
+		return ErrReleaseNotObserved
```
[the remark this answers is not in the document]
Comment (hiddeco) [reply internal/action/verify.go:146]: No, it should be `ErrReleaseNotObserved` as at this point the digest is valid but we can't verify it against the given release object. Given this, we did not observe making the release, as that would have caused the verification to pass.

**Comment by hiddeco** [comment]
Merging this as the CI failure is something that will be addressed by rewriting the actual wiring of the `HelmRelease` reconcilers, and the Go tests pass.

**Pull request merged**